internal/tofu/transform_attach_config_resource.go

```diff
@@ -50,63 +51,23 @@ func (t *AttachResourceConfigTransformer) Transform(g *Graph) error {
 			continue
 		}
-
-		for _, r := range config.Module.ManagedResources {
-			rAddr := r.Addr()
-
-			if rAddr != addr.Resource {
-				// Not the same resource
-				continue
-			}
-
-			log.Printf("[TRACE] AttachResourceConfigTransformer: attaching to %q (%T) config from %s", dag.VertexName(v), v, r.DeclRange)
-			arn.AttachResourceConfig(r)
-
-			// attach the provider_meta info
-			if gnapmc, ok := v.(GraphNodeAttachProviderMetaConfigs); ok {
-				log.Printf("[TRACE] AttachResourceConfigTransformer: attaching provider meta configs to %s", dag.VertexName(v))
-				if config == nil {
-					log.Printf("[TRACE] AttachResourceConfigTransformer: no config set on the transformer for %s", dag.VertexName(v))
-					continue
-				}
-				if config.Module == nil {
-					log.Printf("[TRACE] AttachResourceConfigTransformer: no module in config for %s", dag.VertexName(v))
-					continue
-				}
-				if config.Module.ProviderMetas == nil {
-					log.Printf("[TRACE] AttachResourceConfigTransformer: no provider metas defined for %s", dag.VertexName(v))
-					continue
-				}
-				gnapmc.AttachProviderMetaConfigs(config.Module.ProviderMetas)
-			}
+		var m map[string]*configs.Resource
+		if addr.Resource.Mode == addrs.ManagedResourceMode {
+			m = config.Module.ManagedResources
+		} else if addr.Resource.Mode == addrs.DataResourceMode {
+			m = config.Module.DataResources
 		}
-		for _, r := range config.Module.DataResources {
-			rAddr := r.Addr()
-
-			if rAddr != addr.Resource {
-				// Not the same resource
-				continue
-			}
-
-			log.Printf("[TRACE] AttachResourceConfigTransformer: attaching to %q (%T) config from %#v", dag.VertexName(v), v, r.DeclRange)
-			arn.AttachResourceConfig(r)
-
-			// attach the provider_meta info
-			if gnapmc, ok := v.(GraphNodeAttachProviderMetaConfigs); ok {
-				log.Printf("[TRACE] AttachResourceConfigTransformer: attaching provider meta configs to %s", dag.VertexName(v))
-				if config == nil {
-					log.Printf("[TRACE] AttachResourceConfigTransformer: no config set on the transformer for %s", dag.VertexName(v))
-					continue
-				}
-				if config.Module == nil {
-					log.Printf("[TRACE] AttachResourceConfigTransformer: no module in config for %s", dag.VertexName(v))
-					continue
-				}
-				if config.Module.ProviderMetas == nil {
-					log.Printf("[TRACE] AttachResourceConfigTransformer: no provider metas defined for %s", dag.VertexName(v))
-					continue
+		if m != nil {
+			coord := addr.Resource.String()
+			if r, ok := m[coord]; ok && r.Addr() == addr.Resource {
+				log.Printf("[TRACE] AttachResourceConfigTransformer: attaching to %q (%T) config from %#v", dag.VertexName(v), v, r.DeclRange)
+				arn.AttachResourceConfig(r)
+				if gnapmc, ok := v.(GraphNodeAttachProviderMetaConfigs); ok {
+					log.Printf("[TRACE] AttachResourceConfigTransformer: attaching provider meta configs to %s", dag.VertexName(v))
+					if config.Module != nil && config.Module.ProviderMetas != nil {
```

Comment:
I don't believe `config.Module != nil` can ever be false at this point due to the above access.  It looks like the code this was based on had some extraneous checks.

Reply:
I agree - I tried to keep the original logic as close as possible...

Reply:
This is also missing the "[TRACE] AttachResourceConfigTransformer: no provider metas defined for %s" error message when ProviderMetas is nil.